Hints

StephenVinouze opened the pull request:
I must admit that you did an amazing job. This was exactly what I was looking for, especially the transform into landscape/portrait part of it. I managed to make it work with OpenCV camera but with some drawbacks (landscape only, focus issues, low FPS). Plus, it was not working on my device (Nexus 5X).

This PR intends to fix the `transformToViewCoordinates` method as it did not rotate properly for my device (the preview was correct thanks to `getCameraDisplayOrientation` but not the QRCode points coordinates). It's all good now. I allowed myself to do some code clean but I hope I managed to keep the style/structure intact in order to stick to your work. Also I used your library as a module in the sample instead of using your compiled lib in order to make effective your lib changes while testing the sample.

Finally, I added the ability to set a decode hint map to Zxing `decode`method. Mandatory if your QRCode is encoded with a special charset as it is for my use case.

dlazaro66 commented on qrcodereaderview/src/main/java/com/dlazaro66/qrcodereaderview/QRCodeReaderView.java:
```diff
@@ -101,6 +106,15 @@ public void setQRDecodingEnabled(boolean qrDecodingEnabled) {
+   *
+   * @param decodeHints hints for decoding qrcode
+   */
+  public void setDecodeHints(Map<DecodeHintType, Object> decodeHints) {
```
Can we add `@NonNull` annotation here to the param?

dlazaro66 commented on qrcodereaderview/src/main/java/com/dlazaro66/qrcodereaderview/QRCodeReaderView.java:
```diff
@@ -231,11 +245,11 @@ public void setFrontCamera() {
   @Override public void onPreviewFrame(byte[] data, Camera camera) {
     if (!mQrDecodingEnabled || (decodeFrameTask != null
-        && decodeFrameTask.getStatus() == AsyncTask.Status.RUNNING)) {
+            && decodeFrameTask.getStatus() == AsyncTask.Status.RUNNING)) {
```
I'd prefer to keep the current codestyle. I know is not the best, but I don't want to change it, since I'm planning to adopt the Square codestyle (and change the project style at once) and add it as a constraint to propose a PR

dlazaro66 commented on qrcodereaderview/src/main/java/com/dlazaro66/qrcodereaderview/QRCodeReaderView.java:
```diff
@@ -244,12 +258,10 @@ private boolean checkCameraHardware() {
       return true;
-    } else if (getContext().getPackageManager()
-        .hasSystemFeature(PackageManager.FEATURE_CAMERA_FRONT)) {
+    } else if (getContext().getPackageManager().hasSystemFeature(PackageManager.FEATURE_CAMERA_FRONT)) {
```
same here (about codestyle)

dlazaro66 commented on qrcodereaderview/src/main/java/com/dlazaro66/qrcodereaderview/QRCodeReaderView.java:
```diff
@@ -244,12 +258,10 @@ private boolean checkCameraHardware() {
       return true;
-    } else if (getContext().getPackageManager()
-        .hasSystemFeature(PackageManager.FEATURE_CAMERA_ANY)) {
+    } else if (getContext().getPackageManager().hasSystemFeature(PackageManager.FEATURE_CAMERA_ANY)) {
```
same here (about codestyle)

dlazaro66 commented on qrcodereaderview/src/main/java/com/dlazaro66/qrcodereaderview/QRCodeReaderView.java:
```diff
@@ -262,15 +274,14 @@ private boolean checkCameraHardware() {
     android.hardware.Camera.getCameraInfo(mCameraManager.getPreviewCameraId(), info);
-    WindowManager windowManager =
-        (WindowManager) getContext().getSystemService(Context.WINDOW_SERVICE);
+    WindowManager windowManager = (WindowManager) getContext().getSystemService(Context.WINDOW_SERVICE);
```
same here (about codestyle)

dlazaro66 commented on qrcodereaderview/src/main/java/com/dlazaro66/qrcodereaderview/QRCodeReaderView.java:
```diff
@@ -312,21 +325,12 @@ public DecodeFrameTask(QRCodeReaderView view) {
-      } catch (FormatException e) {
-        Log.d(TAG, "FormatException", e);
+        return view.mQRCodeReader.decode(bitmap, hintsRef.get());
+      } catch (ChecksumException | NotFoundException | FormatException e) {
```
I'd prefer to have different `catch` statements here for every single Exception

dlazaro66 commented on qrcodereaderview/src/main/java/com/dlazaro66/qrcodereaderview/QRCodeReaderView.java:
```diff
@@ -342,70 +346,58 @@ public DecodeFrameTask(QRCodeReaderView view) {
         // Transform resultPoints to View coordinates
-        final PointF[] transformedPoints =
-            transformToViewCoordinates(view, result.getResultPoints());
+        final PointF[] transformedPoints = transformToViewCoordinates(view, result.getResultPoints());
```
same here (about codestyle)

dlazaro66 commented on samples/build.gradle:
```diff
@@ -19,7 +19,8 @@ android {
 
 dependencies {
-  compile 'com.dlazaro66.qrcodereaderview:qrcodereaderview:2.0.1'
+  compile fileTree(dir: 'libs', include: ['*.jar'])
```
I prefer to keep the example working with the latest library dependency. I know you have to change it locally in order to test the current library changes, but for the sake of consistency I strongly prefer to point to the maven release

StephenVinouze commented:
Hi @dlazaro66. I just reverted the indentation to match yours. As of the hint Map, I do believe it should stay that way instead in order to let the user enable/disable hint on the fly. Also, adding this annotation would imply adding the `support-annotations` dependency. I know this is small hindrance but I just wanted to point this out ;)

I hope this suits you!

dlazaro66 commented:
@StephenVinouze Thanks for doing it :)  
>  Also, adding this annotation would imply adding the support-annotations dependency.

Fair enough for me. I'm manually testing it and if everything is ok, I'll merge. Thanks for it :)

dlazaro66 commented:
Hey @StephenVinouze I just checkout your branch, and seems to have a problem while transforming the the QR control points to view coordinates points when the front camera is enabled and you're decoding in landscape configuration. So, said that would you prefer me to merge your PR and change the transformation code or can you revert your changes and we only merge the `Hints` part? Thanks!

StephenVinouze commented:
Hi @dlazaro66, I guess the easiest way would be to start all over by pulling the master and only adding the hints part (which was anyway the point of the PR ^^). That should not be long.

As of your issue with the landscape part with the front camera, I guess this comes from something I changed in the PR as well, unrelated to the hints. I haven't extensively tested my fix hence the issue so good thinking about Unit Tests ;) That will limit regressions.

As of the fix I added, i believe it was due to the issue with Nexus 5X devices that have their back camera upside down... So you should probably investigate how to manage your lib to work in all cases as my fix was not covering every cases. Maybe your Unit Tests will help you :)

dlazaro66 commented:
>  I guess the easiest way would be to start all over by pulling the master and only adding the hints part (which was anyway the point of the PR ^^). That should not be long.

Perfect, that would be great!

> As of the fix I added, i believe it was due to the issue with Nexus 5X devices that have their back camera upside down... So you should probably investigate how to manage your lib to work in all cases as my fix was not covering every cases. Maybe your Unit Tests will help you :)

thanks, definitely, I'll need to figure out how to properly test in other devices, since is hard to generalize the camera behavior for every single device.

Thanks!

StephenVinouze commented:
Yeah, the camera API is a real bitch... To make it easier for me I simply deleted my branch and restarted over by fetching the upstream master and recreated the `hints` branch. However it will probably won't be linked to this PR so we may lose the history of this PR. But I guess it won't be too bad as long as it's merged properly ;)

dlazaro66 commented:
Looks perfect to me!
Thanks!
